components/ai_chat/core/browser/ai_chat_database.cc

```diff
@@ -0,0 +1,914 @@
+#include "components/os_crypt/async/common/encryptor.h"
+#include "sql/statement.h"
+#include "sql/transaction.h"
+
+namespace {
+
+constexpr char kSearchQueriesSeparator[] = "|||";
+
+std::optional<std::string> GetOptionalString(sql::Statement& statement,
+                                             int index) {
+  if (statement.GetColumnType(index) == sql::ColumnType::kNull) {
+    return std::nullopt;
+  }
+  return std::make_optional(statement.ColumnString(index));
+}
+
+}  // namespace
+
+namespace ai_chat {
+
+AIChatDatabase::AIChatDatabase(const base::FilePath& storage_dir,
+                               os_crypt_async::Encryptor encryptor)
+    : db_({.page_size = 4096, .cache_size = 1000}),
+      encryptor_(std::move(encryptor)) {
+  is_initialized_ = Init(storage_dir);
+  if (!is_initialized_) {
+    LOG(ERROR) << "Failed to initialize AI Chat database";
+  }
+}
+
+AIChatDatabase::~AIChatDatabase() = default;
+
+bool AIChatDatabase::IsInitialized() const {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  return is_initialized_;
+}
+
+bool AIChatDatabase::Init(const base::FilePath& db_file_path) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  if (!GetDB().Open(db_file_path)) {
+    return false;
+  }
+
+  sql::Transaction transaction(&GetDB());
+  if (!transaction.Begin()) {
+    return false;
+  }
+
+  if (!CreateConversationTable() || !CreateAssociatedContentTable() ||
+      !CreateConversationEntryTable() || !CreateConversationEntryTextTable() ||
+      !CreateSearchQueriesTable()) {
+    DVLOG(0) << "Failure to create tables";
+    return false;
+  }
+
+  if (!transaction.Commit()) {
+    return false;
+  }
+
+  return true;
+}
+
+std::vector<mojom::ConversationPtr> AIChatDatabase::GetAllConversations() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  static const std::string kQuery =
+      "SELECT conversation.uuid, conversation.title,"
+      " MAX(conversation_entry.date) AS date"
+      " FROM conversation"
+      " LEFT JOIN conversation_entry"
+      " ON conversation_entry.conversation_uuid = conversation.uuid"
+      " GROUP BY conversation.uuid"
+      " ORDER BY date DESC";
+  sql::Statement statement(GetDB().GetCachedStatement(SQL_FROM_HERE, kQuery));
+  CHECK(statement.is_valid());
+
+  std::vector<mojom::ConversationPtr> conversation_list;
+
+  while (statement.Step()) {
+    DVLOG(1) << __func__ << " got a result";
+    mojom::ConversationPtr conversation = mojom::Conversation::New();
+    conversation->uuid = statement.ColumnString(0);
+    conversation->title =
+        DecryptOptionalColumnToString(statement, 1).value_or("");
+    conversation->updated_time = statement.ColumnTime(2);
+    conversation->has_content = true;
+
+    static const std::string kAssociatedContentQuery =
+        "SELECT id, title, url, content_type, content_used_percentage,"
+        " is_content_refined"
+        " FROM associated_content"
+        " WHERE conversation_uuid=?";
```

Comment:
Hmm you might be right as this is an extra query per-Conversation. So once the user reaches 100 conversations then we're doing 101 queries just to get the conversation list. The relationship between `conversation` and `associated_content` is  1 conversation to many associated_content. So if we did a join there would be some duplicated content. I suppose that's not terrible. The advice on the Chromium SQL readme tends to be to keep SELECT queries simple and maintainable and rely on the c++ for any complexity but I agree that a single join isn't too bad (well, a second join). For the conversation_entry SELECT query in `GetConversationEntries` I also don't do JOINs on the events tables, where we have a query for each event type.  But at least there it's a static amount of queries, not one for every row in the first query!

The other, maybe preferable option is to serialize the data to a single table cell on `converation`, since we don't actually cross reference associated content (i.e. it's not a many to many relationship between `conversation` and `associated_content`). It could be an array of data stored in a single cell but then [I'd have to](https://chromium.googlesource.com/chromium/src/+/HEAD/sql/README.md#json) define protobuf message objects.

Reply:
I've converted it to a single query now